boa_engine/src/string.rs

```diff
@@ -312,20 +618,96 @@ impl Inner {
 /// on the stack and a pointer to the data (this is also known as fat pointers).
 /// The `JsString` length and data is stored on the heap. and just an non-null
 /// pointer is kept, so its size is the size of a pointer.
+///
+/// We define some commonly used string constants in the data segment. For these
+/// strings, we no longer allocate memory on the heap to reduce the overhead of
+/// memory allocation and reference counting.
 #[derive(Finalize)]
 pub struct JsString {
-    inner: NonNull<Inner>,
+    inner: Flag,
     _marker: PhantomData<Rc<str>>,
 }
 
+/// It may be an index of [`CONSTANTS_ARRAY`], or a raw pointer of [`Inner`]. Use
+/// the first bit as the flag (Detail: <https://en.wikipedia.org/wiki/Tagged_pointer>).
+///
+/// When the first bit is 0, it represents the address of an [`Inner`]. When the
+/// first bit is 1, it represents an index of [`CONSTANTS_ARRAY`], and the index
+/// number is stored in higher bits.
+///
+/// It uses `NonNull`, which can get benefit from non-null optimization.
```

Comment:
```suggestion
/// This struct uses a technique called tagged pointer to benefit from the fact that newly
/// allocated pointers are always word aligned on 64-bits platforms, making it impossible
/// to have a LSB equal to 1. More details about this technique on the article of Wikipedia
/// about [tagged pointers][tagged_wp].
///
/// # Representation
///
/// If the LSB of the internal [`NonNull<Inner>`] is set (1), then the pointer address represents
/// an index value for [`CONSTANTS_ARRAY`], where the remaining MSBs store the index.
/// Otherwise, the whole pointer represents the address of a heap allocated [`Inner`].
///
/// It uses [`NonNull`], which guarantees that `Flag` (and subsequently [`JsString`])
/// can use the "null pointer optimization" to optimize the size of [`Option<Flag>`].
///
/// # Provenance
///
/// This struct stores a [`NonNull<Inner>`] instead of a [`NonZeroUsize`][std::num::NonZeroUsize]
/// in order to preserve the provenance of our valid heap pointers.
/// On the other hand, all index values are just casted to invalid pointers,
/// because we don't need to preserve the provenance of [`usize`] indices.
///
/// [tagged_wp]: https://en.wikipedia.org/wiki/Tagged_pointer
```